sys/kern/interrupt.c

```diff
@@ -157,30 +183,61 @@ void intr_event_run_handlers(intr_event_t *ie) {
 }
 
 static void intr_thread(void *arg) {
+  intr_thread_t *it = (intr_thread_t *)arg;
+
   while (true) {
-    intr_handler_t *ih;
+    intr_handler_t *ih, *next, *curr;
+
+    ih = NULL;
 
+    /* We look at the list of handlers of our event and search for a handler
+     * marked as IH_DELEGATE. If we don't find any handlers to be serviced, we
+     * go to sleep on our wchan.
+     */
     WITH_INTR_DISABLED {
-      while (TAILQ_EMPTY(&delegated))
-        sleepq_wait(&delegated, NULL);
-      ih = TAILQ_FIRST(&delegated);
-      TAILQ_REMOVE(&delegated, ih, ih_link);
+      TAILQ_FOREACH_SAFE (curr, &it->it_event->ie_handlers, ih_link, next) {
+        if (curr->ih_flags & IH_DELEGATE) {
+          ih = curr;
+          break;
+        }
+      }
+      if (ih == NULL)
+        sleepq_wait(&it->it_wchan, NULL);
     }
 
     ih->ih_service(ih->ih_argument);
 
     intr_event_t *ie = ih->ih_event;
 
     WITH_SPIN_LOCK (&ie->ie_lock) {
-      ie_insert_handler(ie, ih);
-      if (ie->ie_enable)
-        ie->ie_enable(ie);
+
+      /* If the handler is flagged as IH_REMOVE, we remove it similarly to how
+       * we do it in intr_event_remove_handler. We are under a spin lock so we
+       * don't call kfree just yet.
+       */
+      if (ih->ih_flags & IH_REMOVE) {
```

Comment:
This part is common with `intr_event_remove_handler`. Maybe some succinct static function?